### PR TITLE
added search-wrapping for forward and reverse search

### PR DESCRIPTION
**Pull request by unbrand**
Also, the REPEATLAST (for search) scenario is covered so that ctrl-g will wrap if needed.
Note that I chose to ask the user if he wants to wrap. I think that's important because I wanted the user to know if a wrap is taking place. It's an easy hit of the <return> key to initiate the wrap at the point the search has reached the end or top of the file, depending on which direction the search is occurring. 

Props to ne! Adding this behavior was pretty easy because the code is so well thought out. :)

### CODE REVIEW

**Comment by pepa65** [comment]
Cool. The only funny thing I found is that the cursor position always goes one further in the search direction.
And it doesn't seem to work when using Ctrl-G for replacements, which suggests the changes haven't been made at the level of the commands. (Why would REPEAT LAST only wrap for find but not for replace??)

**Comment by vigna** [comment]
There are two issues here:

1) I *hate* this feature. It drives me crazy on Eclipse. Here it is a bit better because there's an explicit confirmation.

2) We must have a way for macros to not stop working (previously, after the last find the action was completed).

**Comment by unbrand** [comment]
Ok, I will look into the cursor position issue you mentioned.
About Ctrl-G, yes, I was treating replacement-wrap as out-of-scope for this fix because I didn't care about replacement-wrap. However, after thinking a bit more, I see your point. I'll take another look to see if there's a good way to allow for replacement-wrapping as well. 

@vigna 1. The thing that bugs me about search-wrapping is when there is no indication that a wrap has occurred. I can't stand that! That's why I ask the user if he wants to do a wrap. The use case is that often, my cursor isn't at the beginning of the file but I just want to start searching for something. If I can easily wrap (and know when it's happening) then for me that's helpful.

2. Hmm, did my code break some macro behavior? If so, I can take a look into it.

**Comment by pepa65** [comment]
I deeply care about the macro functionality. In testing this feature, instead of stopping, it asks for an OK to wrap. This looks good to me. Downside: it doens't stop, but asks for wrapping. Upside: it doesn't continue or give errors.

Maybe the wrapping should be optional... I would probably turn it on by default myself.

**Comment by unbrand** [comment]
@pepa65 Good points. I'll do some testing around macros and I'll think about what it would mean to have the search-wrap behavior be optional. 

Excellent feedback guys, thanks.

**Comment by vigna** [comment]
As I said many times, three were several reasons why I don't want to overfeaturize the editor, in particular not to exaggerate with selectable behavior.

One thing I can imagine is to have a "w" (rather than "y") answer to the find request which would wrap. Certainly modifying an established behavior is a no-no.

**Comment by unbrand** [comment]
@vigna Ok. I see your point. I'll go away and test & think about a good approach keeping in mind what you've said.

**Comment by unbrand** [comment]
Hi guys, I think I've addressed all concerns, and in doing so, simplified my code. I moved the wrapping feature down to the search.c file and that reduced the code quite a bit. Bugs should be squashed now, and the user is asked for a 'w' to wrap if he wants to. REPEAT LAST should work now, as does replace, in both forward and reverse directions.

Please have a look and let me know of any issues. Again, full credit to @vigna for a well-thought-out editor :)

**Comment by unbrand** [comment]
@pepa65 Hmm. When I search for random string, I am asked to wrap, I hit return, and then I see "Not found." in the status bar. Can you provide a bit more detail about what's not working for you?

**Comment by pepa65** [comment]
This is when I do a replace, and then use Repeat Last (Ctrl-G) to replace more, wrap, and the status gets unclear. Maybe you can have a look at what I mean by doing it yourself.

**Comment by unbrand** [comment]
Ah, yes, ok. I can reproduce the scenario you describe. Thanks for the detail. I'll look into it.

**Comment by unbrand** [comment]
Ok,  I fixed the bug that @pepa65 found about the missing "Not found." message when performing a REPEATLAST (ctrl-g). Everything should be working now, according to my testing.

**Comment by vigna** [comment]
@utoddl, maybe we should mumble our ramblings here, so everybody is involved...

**Comment by pepa65** [comment]
Same here, works as intended I would think.

Not that I care much about repeat-replace... But would you perhaps see a chance to implement a REPEAT LAST FIND command (in addition to the REPEAT LAST command that repeats find OR replace, whatever happened last)??

**Comment by unbrand** [comment]
@pepa65 I think this pull-request should stay focused on the wrapping functionality, so I would suggest that an additional REPEAT LAST FIND as you suggest should be submitted separately as a feature request, which I now see as issue #33 I believe. Better to keep it focused there, I think.

**Comment by utoddl** [comment]
I'd like to backup a bit and suggest a less invasive approach which should accomplish what you want with minimal change in work-flow, but without the addition of any new commands, no new prompts, and no change in behaviour wrt macros.

Consider the case in the current/old ne when a search or replace hits the end (either end) of the document. You get a "Not found" message and the search (and macro) stops. (ReplaceAll may report either the number of replacements or "Not Found", but otherwise all these instances are the same case.) The one action we can guarantee is useless at that point is RepeatLast, because there is no matching string to act on on between our current position and the end of the document.

I propose that in this case, and only this case, RepeatLast would affect an otherwise normal repeated operation but "with wrapping". The prior "Not Found" message could be enhanced to include a note to the effect "RepeatLast to wrap" or something succinct but not too wordy to get the point across. There's no prompt you need to satisfy; it's an already available, easily accessible action that would have been useless in this one case that now has a useful effect.

At the risk of falling into the weeds, implementation should be fairly straightforward. The existing "Not Found" circumstances could set a static "repeat_last_wrap" variable in do_action() to 2 (or maybe 3). The top of do_action() would decrement repeat_last_wrap if it's greater than zero. The RepeatLast_A action would somehow indicate to the search.c functions that wrapping should be done once. (I think this part has been implemented in this branch, no?) The search.c functions have the good sense to leave current position in the right place in case of failure, so complications of that sort should be handled already.

**Comment by unbrand** [comment]
@utoddl Thanks for the detailed writeup. At first glance, your suggestions make sense and should cover the most valuable wrapping situations. I certainly understand the desire to be less invasive. No problem. I will sit down and take a more careful look through use cases and code to double-check that no big functionality would be missing with this approach. Thanks again for the considered thinking!

**Comment by unbrand** [comment]
Ok, I've updated the wrap behavior so that it only triggers on a REPEATLAST after a NOT_FOUND. I believe the behavior now matches what you've suggested @utoddl and all the use cases I needed are accounted for. I'll clean up the prefs.c conflict in the pull request.

**Comment by pepa65** [comment]
Love how after a Not Found a Repeat Last makes it wrap!
I do think the message might be confusing for people who don't know what a RepeatLast is...
Maybe no such message is needed.

**Comment by unbrand** [comment]
Yes, it might be slightly confusing for new people, but all the alternatives I can think of aren't as good. I do think a message is needed and following from the suggestion from @utoddl we have "Not found. RepeatLast to wrap." Happy to change, though, if people like something else.

**Comment by unbrand** [comment]
Thanks for the detailed feedback. I see your points. I thought I could keep the changes isolated (mainly) to the search.c file, but I get it now that other non-search actions can intervene which means some logic needs to be outside of search.c. No problem, I'll dig into it. Cheers.

**Comment by utoddl** [comment]
You take criticism with admirable grace.

One thing I usually remember only after a commit is to use the "tabs" syntax highlighter to spot issues with messed up leading and trailing tabs/spaces. Also to look at the code with different tab sizes. We typically use a tab size of 3, but if we're correctly using tabs only for indentation and not for alignment, any tab size should work. (Personally, I detest tabs and never use them, but "When in Project, do what the Projectiles do.")

**Comment by unbrand** [comment]
Heh, criticism is easier to take when it's correct. :) And cheers for being objective with the critique.

About the tabs, yeah, I told my editor to respect the original tabs/spaces, but something must have gone awry. I'll try to make the code more project-consistent. Thanks for the heads-up.

**Comment by unbrand** [comment]
Ok, (most of) search wrapping logic is now moved out of search.c. Please note that for text replacement, wrapping is not enabled. Why? I couldn't make it work in a consistent manner with the functionality of ne. For example, repeatlast would normally trigger a wrap, but when the thing repeated is a replace action, there's a default behavior of "replace one" which I found counter-intuitive (in ne 3.0.1, if you manually put the cursor at the top left position after a replace action triggers a "not found", then ctril-g, you'll see what I mean). Changing that behavior was... not clear as to even if that's desired. So, rather than change existing behavior I left it alone, which meant no wrap for replace. I think it's perfectly ok to not wrap on replace behavior. 

Also, I fixed the one-char-off bug on find-wrap that @utoddl mentioned.

So, from my perspective, this is ready to go, but perhaps there is something I missed in my testing?

**Comment by pepa65** [comment]
This illustrates my case for a separation of Repeat Last and Repeat Last Find. You've implemented wrap for Repeat Last Find. Or, you've broken up the Repeat Last command to behave differently when nothing (more) is found.

Fine with me, I just wish there was a command that would never do a replace, because it makes the whole command/key-combo more dangerous, and you have to "keep state" in your mind... I'd love to just find the next occurrence without running the risk of things getting replaced.

**Comment by unbrand** [comment]
Ah, I see prefs.c has been updated since I forked. No problem, I'll undo my most recent changes to prefs.c.

**Comment by unbrand** [comment]
Ok, this pull request now has the latest from upstream ne. 

@pepa65 After trying to make repeatlast (with search wrap) work with both replace & find I'm starting to lean towards your position. :) I still think it's best to treat it as it's own issue, though, and not part of this pull request.

**Comment by vigna** [comment]
There appears to be several changes that are formatting. Stuff like
    if (cond) 
    {
function arguments on several lines, etc.
For this pull request to be considered, please roll back *everything* that is not a non-whitespace change to the code.

**Comment by unbrand** [comment]
@vigna It looks like I missed some places where the formatting doesn't match the style of the original code. I'll fix it up, no problem.

**Comment by unbrand** [comment]
Ok, everything's cleaned up now. No changes to wrap functionality.
